go/types/simplify_test.go

```diff
@@ -16,141 +16,143 @@ import (
+			{simplifyLists,
+				[]*Type{MakeListType(BoolType)},
+				MakeListType(BoolType)},
+			// list<Number>|list<String>|list<blob> -> list<Number|String|blob>
+			{simplifyLists,
+				[]*Type{MakeListType(NumberType), MakeListType(StringType), MakeListType(BlobType)},
+				MakeListType(MakeUnionType(NumberType, StringType, BlobType))},
+			// list<set<Bool>>|list<set<String>> -> list<set<Bool|String>>
+			{simplifyLists,
+				[]*Type{MakeListType(MakeListType(BoolType)), MakeListType(MakeListType(StringType))},
+				MakeListType(MakeListType(MakeUnionType(BoolType, StringType)))},
+			// list<set<Bool>|list<set<String>>|list<Number> -> list<set<Bool|String>|Number>
+			{simplifyLists,
+				[]*Type{MakeListType(MakeListType(BoolType)), MakeListType(MakeListType(StringType)), MakeListType(NumberType)},
+				MakeListType(MakeUnionType(MakeListType(MakeUnionType(BoolType, StringType)), NumberType))},
+
+			// map<Bool,bool> -> map<Bool,bool>
+			{simplifyMaps,
+				[]*Type{MakeMapType(BoolType, BoolType)},
+				MakeMapType(BoolType, BoolType)},
+			// map<Bool,bool>|map<Bool,string> -> map<Bool,bool|String>
+			{simplifyMaps,
+				[]*Type{MakeMapType(BoolType, BoolType), MakeMapType(BoolType, StringType)},
+				MakeMapType(BoolType, MakeUnionType(BoolType, StringType))},
+			// map<Bool,bool>|map<String,bool> -> map<Bool|String,bool>
+			{simplifyMaps,
+				[]*Type{MakeMapType(BoolType, BoolType), MakeMapType(StringType, BoolType)},
+				MakeMapType(MakeUnionType(BoolType, StringType), BoolType)},
+			// map<Bool,bool>|map<String,string> -> map<Bool|String,bool|String>
+			{simplifyMaps,
+				[]*Type{MakeMapType(BoolType, BoolType), MakeMapType(StringType, StringType)},
+				MakeMapType(MakeUnionType(BoolType, StringType), MakeUnionType(BoolType, StringType))},
+			// map<set<Bool>,bool>|map<set<String>,string> -> map<set<Bool|String>,bool|String>
+			{simplifyMaps,
+				[]*Type{MakeMapType(MakeSetType(BoolType), BoolType), MakeMapType(MakeSetType(StringType), StringType)},
+				MakeMapType(MakeSetType(MakeUnionType(BoolType, StringType)), MakeUnionType(BoolType, StringType))},
+
+			// struct{foo:Bool} -> struct{foo:Bool}
+			{structSimplifier(""),
+				[]*Type{MakeStructTypeFromFields("", FieldMap{"foo": BoolType})},
+				MakeStructTypeFromFields("", FieldMap{"foo": BoolType})},
+			// struct{foo:Bool}|struct{foo:Number} -> struct{foo:Bool|Number}
+			{structSimplifier(""),
+				[]*Type{MakeStructTypeFromFields("", FieldMap{"foo": BoolType}),
+					MakeStructTypeFromFields("", FieldMap{"foo": StringType})},
+				MakeStructTypeFromFields("", FieldMap{"foo": MakeUnionType(BoolType, StringType)})},
+			// struct{foo:Bool}|struct{foo:Bool,bar:Number} -> struct{foo:Bool,bar?:Number}
+			{structSimplifier(""),
+				[]*Type{MakeStructTypeFromFields("", FieldMap{"foo": BoolType}),
+					MakeStructTypeFromFields("", FieldMap{"foo": BoolType, "bar": NumberType})},
+				MakeStructType2("",
+					StructField{"bar", NumberType, !intersectStruct},
```

Comment:
and any place where the expected result type had an optional field it is made required if we are intersecting structs.